libcxx/include/vector

```diff
@@ -1027,6 +1027,10 @@ private:
   }
 
   _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI void __move_assign_alloc(vector&, false_type) _NOEXCEPT {}
+
+  _LIBCPP_CONSTEXPR_SINCE_CXX20 _LIBCPP_HIDE_FROM_ABI pointer __add_alignment_assumption(pointer __p) const _NOEXCEPT {
+    return static_cast<pointer>(__builtin_assume_aligned(__p, alignof(decltype(*__p))));
```

Comment:
It seems that `__builtin_assume_aligned` can't be used in this way in a constexpr context since it returns `void*`:

Example:
```
../../base/containers/checked_iterators_unittest.cc:52:49: error: constexpr variable 'begin' must be initialized by a constant expression
   52 |   constexpr CheckedContiguousConstIterator<int> begin = MakeConstIter(arr, 0u);
      |                                                 ^       ~~~~~~~~~~~~~~~~~~~~~~
../../base/containers/checked_iterators.h:35:12: note: cast from 'void *' is not allowed in a constant expression in C++ standards before C++2c
   35 |     return static_cast<T*>(__builtin_assume_aligned(t, alignof(T)));
      |            ^
```

This works around the compile issue for me in Chromium with clang, though I hope it still optimizes the same.
```cc
  if (std::is_constant_evaluated()) {
    return t;
  } else {
    return static_cast<T*>(__builtin_assume_aligned(t, alignof(T)));
  }
```

I think we'd need to maintain constexpr for these methods in libc++ too, hopefully there are tests that use the iterators from a constexpr context?

Reply:
We should have plenty of `constexpr` tests with iterators in `libcxx/test/std/containers/sequences/vector/iterators.pass.cpp`. Are you able to reduce the issue you were seeing a bit so we can confirm whether there is coverage for this in the test suite?

Reply:
https://godbolt.org/z/ndx8Pe3de shows the error, we can't static_cast from void* in constexpr